Sources/Basics/HTTPClient.swift

```diff
@@ -114,45 +130,65 @@ public struct HTTPClient {
     }
 
     private func _execute(request: Request, requestNumber: Int, observabilityScope: ObservabilityScope?, progress: ProgressHandler?, completion: @escaping CompletionHandler) {
-        if self.shouldCircuitBreak(request: request) {
-            observabilityScope?.emit(warning: "Circuit breaker triggered for \(request.url)")
-            return completion(.failure(HTTPClientError.circuitBreakerTriggered))
+        // wrap completion handler with concurrency control cleanup
+        let originalCompletion = completion
+        let completion: CompletionHandler = { result in
+            // free concurrency control semaphore
+            self.concurrencySemaphore.signal()
+            originalCompletion(result)
         }
 
-        self.underlying(
-            request,
-            { received, expected in
-                if let max = request.options.maximumResponseSizeInBytes {
-                    guard received < max else {
-                        // FIXME: cancel the request?
-                        return completion(.failure(HTTPClientError.responseTooLarge(received)))
-                    }
-                }
-                progress?(received, expected)
-            },
-            { result in
-                switch result {
-                case .failure(let error):
-                    completion(.failure(error))
-                case .success(let response):
-                    // record host errors for circuit breaker
-                    self.recordErrorIfNecessary(response: response, request: request)
-                    // handle retry strategy
-                    if let retryDelay = self.shouldRetry(response: response, request: request, requestNumber: requestNumber) {
-                        observabilityScope?.emit(warning: "\(request.url) failed, retrying in \(retryDelay)")
-                        // TODO: dedicated retry queue?
-                        return self.configuration.callbackQueue.asyncAfter(deadline: .now() + retryDelay) {
-                            self._execute(request: request, requestNumber: requestNumber + 1, observabilityScope: observabilityScope, progress: progress, completion: completion)
+        // we must not block the calling thread (for concurrency control) so nesting this in a queue
+        self.requestsQueue.addOperation {
+            // park the request thread based on the max concurrency allowed
+            self.concurrencySemaphore.wait()
```

Comment:
this diff look terrible but its mostly just nesting under the new queue dispatch calls